lib/index.js

```diff
@@ -0,0 +1,54 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+const {
+  onRouteChannel,
+  onTimeoutChannel,
+  onErrorChannel,
+  onResponseChannel
+} = require('./channels')
+
+function fastifyDiagnostics (fastify, _opts, next) {
+  fastify.addHook('onRoute', (routeOptions) => {
+    onRouteChannel.publish(routeOptions)
+  })
+
+  fastify.addHook('onTimeout', (request, _reply, done) => {
+    onTimeoutChannel.publish({
```

Comment:
Why aren't you publishing `request` and `reply` directly,

Reply:
I feel that we have a lot of properties that don't have good usage through these channels. Publishing both entire objects will only increase the memory propagated. What do you think?

Reply:
>  Publishing both entire objects will only increase the memory propagated. 

diagnostic_channel is not making any copies of those objects.

WDYT @Qard ?

Reply:
If there is no copies involved i'd rather publish `request` and `reply` too.

Reply:
My point is, we don't have clear documentation of all these types inside our `request` and `reply` (I don't remember if have) so far. We have the type definition but not what each property means. I would not expose properties without clear documentation.

Reply:
Another that I think is:

Is not a security issue expose all the request/reply data? I mean, if some evil module subscribes to `onRequest` and gets the `user` and `password` for an auth request or even makes a monkey-patch of reply functions for instance since we send the real object of request/reply.
(I'm not 100% sure about what I said, correct me if something is wrong please)

Reply:
> My point is, we don't have clear documentation of all these types inside our request and reply (I don't remember if have) so far. We have the type definition but not what each property means. I would not expose properties without clear documentation.

I don't understand this. We have massive docs available on the web as well as typescript support for those.

> Is not a security issue expose all the request/reply data? I mean, if some evil module subscribes to onRequest and gets the user and password for an auth request or even makes a monkey-patch of reply functions for instance since we send the real object of request/reply.
(I'm not 100% sure about what I said, correct me if something is wrong please)

If an attacker can run JS on your VM, you are deeply in trouble and can have access to your full memory space.

Reply:
> If an attacker can run JS on your VM, you are deeply in trouble and can have access to your full memory space.

Fair point.

Indeed. My mistake on Fastify `reply/request`, I just found the type docs and don't found the properties documentation and I completely forgot of: [reply docs](https://www.fastify.io/docs/latest/Reply/#reply) and [request docs](https://www.fastify.io/docs/latest/Request/).

I'll update the objects.